packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -0,0 +1,519 @@
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../aux/access/ISubmitBlocksCallback.sol";
+import "../thirdparty/BytesUtil.sol";
+import "../core/iface/ExchangeData.sol";
+import "../core/iface/IExchangeV3.sol";
+import "../lib/AddressUtil.sol";
+import "../lib/ERC20SafeTransfer.sol";
+import "../lib/MathUint.sol";
+
+import "../core/impl/libtransactions/AmmUpdateTransaction.sol";
+import "../core/impl/libtransactions/DepositTransaction.sol";
+import "../core/impl/libtransactions/WithdrawTransaction.sol";
+
+contract AmmPool is ISubmitBlocksCallback {
+
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using BytesUtil         for bytes;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+
+    event Deposit(
+        address  owner,
+        uint96[] amounts
+    );
+
+    event Withdrawal(
+        address  owner,
+        uint96[] amounts
+    );
+
+    event JoinPoolRequested(
+        address  owner,
+        uint     poolAmountOut,
+        uint96[] maxAmountsIn
+    );
+
+    event ExitPoolRequested(
+        address  owner,
+        uint     poolAmountIn,
+        uint96[] minAmountsOut
+    );
+
+    uint public constant MAX_AGE_REQUEST_UNTIL_POOL_SHUTDOWN = 7 days;
+
+    IExchangeV3 public exchange;
+    uint32      public accountID;
+
+    uint        public shutdownTimestamp = 0;
+
+    struct PoolJoin
+    {
+        uint poolAmountOut;
+        uint96[] maxAmountsIn;
+    }
+
+    struct PoolExit
+    {
+        uint poolAmountIn;
+        uint96[] minAmountsOut;
+    }
+
+    enum WorkType
+    {
+        JOIN,
+        EXIT
+    }
+
+    enum Status
+    {
+        PENDING,
+        SUCCESS,
+        FAILED
+    }
+
+    struct QueueItem
+    {
+        address  owner;
+        uint64   timestamp;
+        Status   status;
+        WorkType workType;
+        PoolJoin join;
+        PoolExit exit;
+    }
+
+    struct Token
+    {
+        address addr;
+        uint96  weight;
+        uint16  tokenID;
+    }
+
+    uint8 public feeBips;
+    Token[] public tokens;
+
+    QueueItem[] public queue;
+    uint public queuePos = 0;
+
+    // A map from an owner to a token to the balance
+    mapping (address => mapping (address => uint)) balance;
+    // A map from an owner to a token to the balance locked
+    mapping (address => mapping (address => uint)) locked;
+    // A map a token to the total balance owned by different users (NOT the balance of the pool itself)
+    mapping (address => uint) totalBalance;
+
+    modifier onlyExchangeOwner()
+    {
+        require(msg.sender == exchange.owner(), "UNAUTHORIZED");
+        _;
+    }
+
+    modifier online()
+    {
+        require(isOnline(), "NOT_ONLINE");
+        _;
+    }
+
+    modifier offline()
+    {
+        require(!isOnline(), "NOT_OFFLINE");
+        _;
+    }
+
+    function setupPool(
```

Comment:
Depends on if we want pools to be upgradeable or not.

But just that we're on the same page of this pool implementation: for now this is just for testing (this is why I added the contract to the test folder). While some of this code will definitely be used in some form, it'll depend on how we go forward on the actual pool implementations, otherwise some of this code will simply be used as the glue between Loopring and e.g. the balancer pool contracts.